File: classes/AVLTree.cpp
#include "AVLTree.h"

// Constructor
template <typename T>
AVLTree<T>::AVLTree() : root(nullptr) {}

// Height Calculation
template <typename T>
int AVLTree<T>::height(Node* n) {
    return n ? n->height : 0;
}

// Balance Factor Calculation
template <typename T>
int AVLTree<T>::getBalance(Node* n) {
    return n ? height(n->left) - height(n->right) : 0;
}

// Right Rotation
template <typename T>
typename AVLTree<T>::Node* AVLTree<T>::rightRotate(Node* y) {
    Node* x = y->left;
    Node* T2 = x->right;

    x->right = y;
    y->left = T2;

    y->height = std::max(height(y->left), height(y->right)) + 1;
    x->height = std::max(height(x->left), height(x->right)) + 1;

    return x;
}

// Left Rotation
template <typename T>
typename AVLTree<T>::Node* AVLTree<T>::leftRotate(Node* x) {
    Node* y = x->right;
    Node* T2 = y->left;

    y->left = x;
    x->right = T2;

    x->height = std::max(height(x->left), height(x->right)) + 1;
    y->height = std::max(height(y->left), height(y->right)) + 1;

    return y;
}

// Insertion
template <typename T>
typename AVLTree<T>::Node* AVLTree<T>::insert(Node* node, T key) {
    if (!node)
        return new Node(key);

    if (key < node->key)
        node->left = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);
    else
        return node;

    node->height = 1 + std::max(height(node->left), height(node->right));

    int balance = getBalance(node);

    if (balance > 1 && key < node->left->key)
        return rightRotate(node);

    if (balance < -1 && key > node->right->key)
        return leftRotate(node);

    if (balance > 1 && key > node->left->key) {
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }

    if (balance < -1 && key < node->right->key) {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }

    return node;
}

// Public Insert Method
template <typename T>
void AVLTree<T>::insert(T key) {
    root = insert(root, key);
}

// PreOrder Traversal
template <typename T>
void AVLTree<T>::preOrder(Node* root) {
    if (root) {
        std::cout << root->key << " ";
        preOrder(root->left);
        preOrder(root->right);
    }
}

// Public PreOrder Method
template <typename T>
void AVLTree<T>::preOrder() {
    preOrder(root);
}

File: classes/AVLTree.js
class AVLNode {
    constructor(key, value) {
        this.key = key;
        this.value = new Set();
        this.value.add(value);
        this.left = null;
        this.right = null;
        this.height = 1;
        this.phonetic = false;
    }
}

class AVLTree {
    constructor() {
        this.root = null;
    }

    getHeight(node) {
        if (!node) return 0;
        return node.height;
    }

    updateHeight(node) {
        node.height = Math.max(this.getHeight(node.left), this.getHeight(node.right)) + 1;
    }

    getBalanceFactor(node) {
        if (!node) return 0;
        return this.getHeight(node.left) - this.getHeight(node.right);
    }

    rotateRight(y) {
        const x = y.left;
        const temp = x.right;

        x.right = y;
        y.left = temp;
        this.updateHeight(y);
        this.updateHeight(x);

        return x;
    }

    rotateLeft(x) {
        const y = x.right;
        const temp = y.left;

        y.left = x;
        x.right = temp;
        this.updateHeight(x);
        this.updateHeight(y);

        return y;
    }

    balance(node) {
        if (!node) return null;

        this.updateHeight(node);
        const balanceFactor = this.getBalanceFactor(node);

        if (balanceFactor > 1) {
            if (this.getBalanceFactor(node.left) >= 0) {
                return this.rotateRight(node);
            }
            else {
                node.left = this.rotateLeft(node.left);
                return this.rotateRight(node);
            }
        }
        else if (balanceFactor < -1) {
            if (this.getBalanceFactor(node.right) <= 0) {
                return this.rotateLeft(node);
            }
            else {
                node.right = this.rotateRight(node.right);
                return this.rotateLeft(node);
            }
        }

        return node;
    }

    insert(key, value) {
        this.root = this._insert(this.root, key, value);
    }

    _insert(node, key, value) {
        if (!node) return new AVLNode(key, value);

        if (key < node.key) {
            node.left = this._insert(node.left, key, value);
        } else if (key > node.key) {
            node.right = this._insert(node.right, key, value);
        } else {
            node.value.add(value);
            return node;
        }
        this.updateHeight(node);

        return this.balance(node);
    }

    search(key) {
        let output;
        if (this.contains(key)) {
            output = this._search(this.root, key).value;
            if (this.phonetic) output = new Set([...output].map(this.toPhonetic));
        } else {
            output = new Set().add("[No Brief Found]");
        }

        return output;
    }

    _search(node, key) {
        if (!node) return null;
        if (key < node.key) {
            return this._search(node.left, key);
        } else if (key > node.key) {
            return this._search(node.right, key);
        } else {
            return node;
        }
    }

    contains(key) {
        return this._search(this.root, key) !== null;
    }

    toPhonetic(word) {
        if (typeof word !== 'string') {
            console.error('Input must be a string');
            return word;
        }
        const replacements = {
            "TPH": "N",
            "TP": "F",
            "KP": "X",
            "HR": "L",
            "TK": "D",
            "PW": "B",
            "EU": "I",
            "SR": "V",
            "TKPW": "G",
            "BG": "K",
            "PH": "M",
            "SKWR": "J",
            "KWR": "Y",
            "PL": "M",
            "PB": "N",
            "PBLG": "J",
            "KH": "CH",
            "BGS": "CTION",
            "GS": "TION",
            "FP": "CH",
            "RB": "SH",
            "FRB": "RV",
            "AOU": "UU",
            "AEU": "AI",
            "OEU": "OI",
            "AOE": "EE",
            "AOEU": "II"
        };

        let output = word;
        for (const [sequence, replacement] of Object.entries(replacements)) {
            const regex = new RegExp(sequence, 'g');
            output = output.replace(regex, replacement);
        }

        return output;
    }
}


module.exports = AVLTree;